### ISSUE TITLE
LogWarnings in photonID 2016 V2 (as if ID is not POG approved)

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#19799]
these are present since 930pre1
and are related to #18645

```
%MSG-w IdInformation:  VersionedPhotonIdProducer:egmPhotonIDs@ctor  18-Jul-2017 18:30:59 CEST pre-events
Instantiated ID: cutBasedPhotonID-Spring16-V2p2-loose
with MD5 hash: d6ce6a4f3476294bf0a3261e00170daf
This ID is not POG approved and likely under development!!!
Please make sure to report your progress with this ID at the next relevant POG meeting.

%MSG
%MSG-w IdInformation:  VersionedPhotonIdProducer:egmPhotonIDs@ctor  18-Jul-2017 18:30:59 CEST pre-events
Instantiated ID: cutBasedPhotonID-Spring16-V2p2-medium
with MD5 hash: c739cfd0b6287b8586da187c06d4053f
This ID is not POG approved and likely under development!!!
Please make sure to report your progress with this ID at the next relevant POG meeting.

%MSG
%MSG-w IdInformation:  VersionedPhotonIdProducer:egmPhotonIDs@ctor  18-Jul-2017 18:30:59 CEST pre-events
Instantiated ID: cutBasedPhotonID-Spring16-V2p2-tight
with MD5 hash: bdb623bdb1a15c13545020a919dd9530
This ID is not POG approved and likely under development!!!
Please make sure to report your progress with this ID at the next relevant POG meeting.

```

@ikrav @jainshilpi @rafaellopesdesa

**Comment by slava77** [comment]
assign reconstruction

**Comment by ikrav** [comment]
Hi Slava,

   I have created a 930pre1 test area, ran a standard test with the “offending” photon IDs,
and have not observed the issue.

However, I suspect I know what this issue is about. There is a check of the MD5 sum of all
IDs in the VID framework. The check goes wrong under certain circumstances, as if MD5 sum
itself is computed differently when the code is run on different machines. It was an occasional
issue on lxplus in the past, and may be related to where the files physically located at (afs vs
elsewhere). On some lxplus nodes one would get this warning, while on others one wouldn’t. 
Then sometimes one would get these warnings, but after bringing the package to local directory
with “git cms-addpkg RecoEgamma/PhotonIdentification”, the warnings would go away.

Since Lindsey wrote all the md5sum handling machinery, I turned to Lindsey back when we had
a surge of complaints, in Jan-Feb of 2016. Lindsey said he will investigate it, but it would take a bit
of time… He seemed to have a general understanding of the issue, and it was not trivial.

Meanwhile, I thought Lindsey silenced the complaint by adding lines like these in all our
ID definition files through lines like these:

https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/PhotonIdentification/python/Identification/cutBasedPhotonID_Spring16_V2p2_cff.py#L154-L156

I wonder if this is not working anymore, for some reason.

The other possibility was that someone perhaps changed the .py files
and the MD5 sum is out of date, but I checked and this doesn’t seem to be the case.

I am adding @lgray and ccing to ID conveners.

Ilya

**Comment by ikrav** [comment]
Is there a simple way to reproduce this problem? Because it seems
that Lindsey’s silencing this issue, implemented in 

https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/SelectorUtils/interface/VersionedIdProducer.h#L102-L110

and with the setting from the ID

https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/PhotonIdentification/python/Identification/cutBasedPhotonID_Spring16_V2p2_cff.py#L154-L156

should work, but somehow it does not seem to.

Thanks,
ILya

**Comment by davidlange6** [comment]
runTheMatrix.py -l 10024

and look at the step3 log file.

**Comment by ikrav** [comment]
Thanks David, I can reproduce the problem now. 
Ilya

**Comment by ikrav** [comment]
I have made some progress in understanding these warnings. Apparently, a new
place where cut-based photon ID is used has been introduced very recently (mid-June by @Sam-Harper ?)
in DQM Offline/HLT. It is here:

https://github.com/cms-sw/cmssw/blob/master/DQMOffline/Trigger/python/HLTEGTnPMonitor_cfi.py#L777-L793
and is used from 
https://github.com/cms-sw/cmssw/blob/master/DQMOffline/Trigger/python/EgammaMonitoring_cff.py#L3-L8

Now cut-based photon IDs are imported twice. One of the time there is no warning,
the other time there is warning. The warning comes because the field isPOGApproved
in the ParameterSet of the ID (such as "cutBasedPhotonID-Spring16-V2p2-tight”) is set to
“false” in one of the two instances, and “true” in the other. How this can come to be, I do not
understand at the moment, since the source cff of this ParameterSet has at the end an unconditional
setting of this parameter to true. Some sort of interference of two instances? Something is set
for VID in a non-intended way in the new addition in DQMOffline? This requires more of looking around.
I am not sure if Sam is available to look at his additions, though, at this time.

This new ID was introduced in 9_2_2, and there there were no warnings. I see that
the DQMOffline new piece is already in 9_2_5, but was not there in 9_2_4.

Ilya
PS in the end, this does not appear to be the problem of the MD5 sums that 
I originally mentioned.

**Comment by slava77** [comment]
On 7/19/17 9:15 AM, ikrav wrote:
> Now cut-based photon IDs are imported twice. One of the time there is no
> warning,
> the other time there is warning. The warning comes because the field
> isPOGApproved
> in the ParameterSet of the ID (such as
> "cutBasedPhotonID-Spring16-V2p2-tight”) is set to
> “false” in one of the two instances, and “true” in the other. How this
> can come to be, I do not
> understand at the moment, since the source cff of this ParameterSet has
> at the end an unconditional
> setting of this parameter to true. Some sort of interference of two
> instances? Something is set
> for VID in a non-intended way in the new addition in DQMOffline? This
> requires more of looking around.
> I am not sure if Sam is available to look at his additions, though, at
> this time.

does this affect only this isPOGApproved parameter or do other values in
the VID PSet get reset/changed?
The latter could be a more significant problem.

**Comment by ikrav** [comment]
I have called the dump() on the ParameterSet that defines all cut types and
cut values from the VersionedIdProducer.h constructor, and then have done 
a diff, and there is no difference between the two instances except for the 
isPOGApproved field.

Ilya

**Comment by Dr15Jones** [comment]
How is the ParameterSet obtained? Is it the one passed into the module's constructor or is it pulled directly from the ParameterSetRegistry?

**Comment by ikrav** [comment]
Let me try to answer this to the best of my understanding.

The warnings about not-POG-approved state come from the constructor
of the VersionedIdProducer. The list of parameter sets for the IDs is pulled
out of the parameter set that is passed into the constructor:

https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/SelectorUtils/interface/VersionedIdProducer.h#L61
(the warning is printed from the code about a page down in that constructor).

However, I do not know if there are multiple instances of theVersionedIdProducer,
since VID is invoked both from miniAOD_tools.py in this example (step3) for miniAOD 
production as well as from the HLTEGTnPMonitor_cfi.py (I have not spend enough
looking at this to understand, but I think two separate sequences are defined somewhere).

Regards,
ILya

**Comment by Sam-Harper** [comment]
Hey everybody. A current trigger crisis (whats new?) is preventing me from really commenting. 

Its a completely separate sequence I put i because I needed VID on AOD objects, not miniAOD objects.  And I must have screwed it up somehow. Sorry about that that.  

One is running on miniAOD input, one is running on AOD input. I admit, it could be better. Not sure why the switch wasnt flipped. 

Unfortunately I cant be of that much more help than that as there are other fires much more serious than this about.

**Comment by Sam-Harper** [comment]
So HLT crisis over. I had a look. Oh fun.
 
The problem is that VID deletes the parameter from the pset of the ID during setup.
As Slava is no doubt painfully aware my python skills are weak but to me it's the same PSet we pass in twice (as in the same one, a reference not a copy). 
https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/SelectorUtils/python/tools/vid_id_tools.py#L18-L20

I have two VID modules is a bit of a hack because I needed AOD value maps when RECO needed miniAOD keyed valuemaps.

The problem is that when you have two VID modules calling the same ID module, the second time the isPOGApproved is deleted from the pset (because VID doesnt allow untracked parameters and wont run). Because the pset the second time around doesnt have this, it defaults to pog approved. It didnt happen to the electrons as used the summer16 rather than spring16 values (why reco isnt using summer16 is a story for another day). 

@lgray. this is horrid. 

A solution has to be found
1) I need VID value maps keyed to AOD objects as the DQM group tells me it must run on RECO
2) miniAOD needs miniAOD keyed valuemaps for VID.
3) creating two instances of VID cause problems such as this

**Comment by lgray** [comment]
Just read through all of this.

OK probably the easiest thing to do is just allow exactly one untracked
parameter (the approved flag).
Then no deleting is required and it doesn't change any MD5 sums.

Best,
-Lindsey

**Comment by Sam-Harper** [comment]
Thanks Lindsey. Btw Ilya asked me privately, why dont we just make the pog approved tracked? Given the hacks (no judgement, we've all been there, there are times for elegance and there are times for getting it working now), I assumed that there a deep reason why not to.

**Comment by Sam-Harper** [comment]
Also there main reason I tagged you, is there any deep conceptual reason why two copies of VID running over the same ID would be bad? I assume not but just want to be sure.

**Comment by lgray** [comment]
no, not really any reason it should be bad. It was just thought of for
analysis, where that use didn't make any sense.

switching it to tracked or untracked would cause you to need to recalculate
the md5s every time you changed it, kind of annoying.

-L

**Comment by Sam-Harper** [comment]
Fair enough. Thanks for the clarification. And yes, thats a good reason why it was untracked, knew there had to be one :)

**Comment by slava77** [comment]
+1

fixed in #19872
